File: src/main/javabak/tile_entities/capabilities/ProcessedItemStackHandler.java
package redcrafter07.processed.block.tile_entities.capabilities;

import net.minecraft.core.HolderLookup;
import net.minecraft.core.NonNullList;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.Tag;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

public class ProcessedItemStackHandler extends IProcessedItemHandler<CompoundTag> {
    protected NonNullList<ItemStack> items;

    public ProcessedItemStackHandler(NonNullList<ItemStack> items) {
        this.items = items;
    }

    public ProcessedItemStackHandler(int size) {
        this(NonNullList.withSize(size, ItemStack.EMPTY));
    }

    public ProcessedItemStackHandler() {
        this(1);
    }

    @Override
    public int getSlots() {
        return items.size();
    }

    @Override
    public ItemStack getStackInSlot(int slot) {
        if (slot >= items.size()) return ItemStack.EMPTY;
        return items.get(slot);
    }

    @Override
    public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
        if (stack.isEmpty()) return ItemStack.EMPTY;
        if (!isItemValid(slot, stack)) return stack;
        if (invalidSlotIndex(slot)) return ItemStack.EMPTY;
        final var existing = items.get(slot);
        var limit = getStackLimit(slot, stack);
        if (!existing.isEmpty()) {
            if (!ItemStack.isSameItemSameComponents(stack, existing))
                return stack;

            limit -= existing.getCount();
        }

        if (limit <= 0) return stack;
        final var reachedLimit = stack.getCount() > limit;

        if (!simulate) {
            if (existing.isEmpty()) {
                items.set(slot, (reachedLimit) ? stack.copyWithCount(limit) : stack);
            } else {
                existing.grow((reachedLimit) ? limit : stack.getCount());
            }
            setChanged(slot);
        }

        return (reachedLimit) ? stack.copyWithCount(stack.getCount() - limit) : ItemStack.EMPTY;
    }

    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        if (amount <= 0) return ItemStack.EMPTY;
        if (invalidSlotIndex(slot)) return ItemStack.EMPTY;
        final var existing = items.get(slot);
        if (existing.isEmpty()) return ItemStack.EMPTY;
        final var toExtract = Math.min(amount, existing.getMaxStackSize());
        if (existing.getCount() <= toExtract) {
            if (!simulate) {
                items.set(slot, ItemStack.EMPTY);
                setChanged(slot);
                return existing;
            } else {
                return existing.copy();
            }
        } else {
            if (!simulate) {
                items.set(slot, existing.copyWithCount(existing.getCount() - toExtract));
                setChanged(slot);
            }

            return existing.copyWithCount(toExtract);
        }
    }

    @Override
    public int getSlotLimit(int slot) {
        if (invalidSlotIndex(slot)) return 0;
        if (items.get(slot).isEmpty()) return Item.ABSOLUTE_MAX_STACK_SIZE;
        return Math.min(Item.ABSOLUTE_MAX_STACK_SIZE, items.get(slot).getMaxStackSize());
    }

    @Override
    public boolean isItemValid(int slot, ItemStack stack) {
        return true;
    }

    @Override
    public void setStackInSlot(int slot, ItemStack stack) {
        if (invalidSlotIndex(slot)) return;
        items.set(slot, stack);
        setChanged(slot);
    }

    @Override
    public @NotNull CompoundTag serializeNBT(HolderLookup.Provider provider) {
        final var nbtTagList = new ListTag();
        for (int i = 0; i < items.size(); ++i) {
            if (items.get(i).isEmpty()) continue;
            final var itemTag = new CompoundTag();
            itemTag.putInt("Slot", i);
            nbtTagList.add(items.get(i).save(provider, itemTag));
        }
        final var nbt = new CompoundTag();
        nbt.put("Items", nbtTagList);
        nbt.putInt("Size", items.size());
        return nbt;
    }

    @Override
    public void deserializeNBT(HolderLookup.Provider provider, CompoundTag nbt) {
        setSize((nbt.contains("Size", Tag.TAG_INT)) ? nbt.getInt("Size") : items.size());
        final var tagList = nbt.getList("Items", Tag.TAG_COMPOUND);
        for (int i = 0; i < tagList.size(); ++i) {
            final var itemTags = tagList.getCompound(i);
            var slot = itemTags.getInt("Slot");

            if (slot >= 0 && slot < items.size())
                items.set(slot, ItemStack.parseOptional(provider, itemTags));

        }
    }

    protected void setSize(int newSize) {
        items = NonNullList.withSize(newSize, ItemStack.EMPTY);
        setChanged(-1);
    }

    protected boolean invalidSlotIndex(int slot) {
        return (slot < 0 || slot >= items.size());
    }

    protected int getStackLimit(int slot, ItemStack stack) {
        return Math.min(getSlotLimit(slot), stack.getMaxStackSize());
    }
}

File: src/main/kotlin/redcrafter07/processed/block/tile_entities/capabilities/ProcessedItemStackHandler.kt
package redcrafter07.processed.block.tile_entities.capabilities

import net.minecraft.core.HolderLookup
import net.minecraft.core.NonNullList
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.nbt.Tag
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import kotlin.math.min

open class ProcessedItemStackHandler(protected var items: NonNullList<ItemStack>): ProcessedItemHandler<CompoundTag>() {
    constructor(size: Int) : this(NonNullList.withSize(size, ItemStack.EMPTY))
    constructor() : this(1)

    override fun getSlots(): Int {
        return items.size
    }

    override fun getStackInSlot(slot: Int): ItemStack {
        if (slot >= items.size) return ItemStack.EMPTY
        return items[slot]
    }

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        if (stack.isEmpty) return ItemStack.EMPTY
        if (!isItemValid(slot, stack)) return stack
        if (invalidSlotIndex(slot)) return ItemStack.EMPTY
        val existing = items[slot]
        var limit = getStackLimit(slot, stack)
        if (!existing.isEmpty) {
            if (!ItemStack.isSameItemSameComponents(stack, existing)) return stack

            limit -= existing.count
        }

        if (limit <= 0) return stack
        val reachedLimit = stack.count > limit

        if (!simulate) {
            if (existing.isEmpty) {
                items[slot] = if (reachedLimit) stack.copyWithCount(limit) else stack
            } else {
                existing.grow(if (reachedLimit) limit else stack.count)
            }
            setChanged(slot)
        }

        return if (reachedLimit) stack.copyWithCount(stack.count - limit) else ItemStack.EMPTY
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        if (amount <= 0) return ItemStack.EMPTY
        if (invalidSlotIndex(slot)) return ItemStack.EMPTY
        val existing = items[slot]
        if (existing.isEmpty) return ItemStack.EMPTY
        val toExtract = min(amount, existing.maxStackSize)
        if (existing.count <= toExtract) {
            if (!simulate) {
                items[slot] = ItemStack.EMPTY
                setChanged(slot)
                return existing
            } else {
                return existing.copy()
            }
        } else {
            if (!simulate) {
                items[slot] = existing.copyWithCount(existing.count - toExtract)
                setChanged(slot)
            }

            return existing.copyWithCount(toExtract)
        }
    }

    override fun getSlotLimit(slot: Int): Int {
        if (invalidSlotIndex(slot)) return 0
        if (items[slot].isEmpty) return Item.ABSOLUTE_MAX_STACK_SIZE
        return min(Item.ABSOLUTE_MAX_STACK_SIZE, items[slot].maxStackSize)
    }

    override fun isItemValid(slot: Int, stack: ItemStack): Boolean {
        return true
    }

    override fun setStackInSlot(slot: Int, stack: ItemStack) {
        if (invalidSlotIndex(slot)) return
        items[slot] = stack
        setChanged(slot)
    }

    override fun serializeNBT(provider: HolderLookup.Provider): CompoundTag {
        val nbtTagList = ListTag()
        for (i in items.indices) {
            if (items[i].isEmpty) continue
            val itemTag = CompoundTag()
            itemTag.putInt("Slot", i)
            nbtTagList.add(items[i].save(provider, itemTag))
        }
        val nbt = CompoundTag()
        nbt.put("Items", nbtTagList)
        nbt.putInt("Size", items.size)
        return nbt
    }

    override fun deserializeNBT(provider: HolderLookup.Provider, nbt: CompoundTag) {
        setSize(if (nbt.contains("Size", Tag.TAG_INT.toInt())) nbt.getInt("Size") else items.size)
        val tagList = nbt.getList("Items", Tag.TAG_COMPOUND.toInt())
        for (i in tagList.indices) {
            val itemTags = tagList.getCompound(i)
            val slot = itemTags.getInt("Slot")

            if (slot >= 0 && slot < items.size) items[slot] = ItemStack.parseOptional(provider, itemTags)
        }
    }

    protected fun setSize(newSize: Int) {
        items = NonNullList.withSize(newSize, ItemStack.EMPTY)
        setChanged(-1)
    }

    protected fun invalidSlotIndex(slot: Int): Boolean {
        return (slot < 0 || slot >= items.size)
    }

    protected fun getStackLimit(slot: Int, stack: ItemStack): Int {
        return min(getSlotLimit(slot), stack.maxStackSize)
    }
}